crates/rome_js_analyze/src/semantic_analyzers/nursery/no_void_elements_with_children.rs

```diff
@@ -0,0 +1,193 @@
+use crate::react::is_react_create_element;
+use crate::semantic_services::Semantic;
+use rome_analyze::context::RuleContext;
+use rome_analyze::{declare_rule, Rule, RuleCategory, RuleDiagnostic};
+use rome_console::{markup, MarkupBuf};
+use rome_js_syntax::{JsCallExpression, JsxElement, JsxSelfClosingElement};
+use rome_rowan::{declare_node_union, AstNode, AstNodeList};
+
+declare_rule! {
+    /// This rules prevents void elements (AKA self-closing elements) from having children.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// <br>invalid child</br>
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// <img alt="some text" children={"some child"} />
+    /// ```
+    pub(crate) NoVoidElementsWithChildren {
+        version: "0.10.0",
+        name: "noVoidElementsWithChildren",
+        recommended: false,
+    }
+}
+
+declare_node_union! {
+    pub(crate) NoVoidElementsWithChildrenQuery = JsxElement | JsCallExpression | JsxSelfClosingElement
+}
+
+const VOID_DOM_ELEMENTS: [&str; 16] = [
+    "area", "base", "br", "col", "embed", "hr", "img", "input", "keygen", "link", "menuitem",
+    "meta", "param", "source", "track", "wbr",
+];
+
+pub(crate) struct NoVoidElementsWithChildrenState {
+    /// The name of the element that triggered the rule
+    element_name: String,
+    /// If the current element has children props
+    children_cause: bool,
+    /// If the current element has the prop `dangerouslySetInnerHTML`
+    dangerous_prop_case: bool,
+}
+
+impl NoVoidElementsWithChildrenState {
+    fn new(element_name: impl Into<String>) -> Self {
+        Self {
+            element_name: element_name.into(),
+            children_cause: false,
+            dangerous_prop_case: false,
+        }
+    }
+
+    fn with_children_cause(&mut self, cause: bool) {
+        self.children_cause = cause;
+    }
+
+    fn with_dangerous_prop_cause(&mut self, cause: bool) {
+        self.dangerous_prop_case = cause;
+    }
+
+    fn message(&self) -> MarkupBuf {
+        match (self.children_cause, self.dangerous_prop_case) {
+            (true, true) => {
+                (markup! {
+                    <Emphasis>{self.element_name}</Emphasis>" is a void element tag and must not have the "<Emphasis>"children"</Emphasis>
+                    " or the"<Emphasis>"dangerouslySetInnerHTML"</Emphasis>" prop."
```

Comment:
The markup is missing a space here:
```suggestion
                    " or the "<Emphasis>"dangerouslySetInnerHTML"</Emphasis>" prop."
```